compiler/il/OMRNode.cpp

```diff
@@ -2316,17 +2316,115 @@ OMR::Node::dontEliminateStores(bool isForLocalDeadStore)
 bool
 OMR::Node::isNotCollected()
    {
-   TR::Compilation * comp = TR::comp();
-   if (self()->getOpCode().hasSymbolReference() && self()->getSymbolReference()->getSymbol()->isNotCollected())
-      return true;
-   else if (comp->generateArraylets() &&
-            (self()->getOpCode().isArrayRef() &&
-             (self()->getFirstChild()->isNotCollected() || !self()->getFirstChild()->isInternalPointer())))
-      return true;
-   return false;
+   return !self()->computeIsCollectedReference();
    }
 
 
+/**
+ * \brief
+ *    Check if the node is an internal pointer.
+ *
+ * \return
+ *    Return true if the node is an internal pointer, false otherwise.
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
```

Comment:
I believe that `aiadd`, `aiuadd`, `aladd`, and `aluadd` are the only opcodes this function is currently meant to handle. The comment likely means to say that it only works for nodes whose opcodes satisfy `hasPinningArrayPointer()`, which is perhaps poorly named. Callers check `hasPinningArrayPointer()` before calling this function.

Originally the result for any other opcode would have been false. Precisely because we might later want to return true for loads of internal pointer temps, or possibly in other reasonable cases, Liqun documented this opcode restriction and added the assertion below to check that it holds. This way, callers can't assume that a true result informs them about the opcode of the node, and we're free to relax the opcode restriction in the future.

Since then, the implementing `computeIsCollectedReference()` has changed so that it actually will return true for loads of internal pointer temps ([see here](https://github.com/eclipse/omr/pull/1178/files/ade5a1ccac7694304d9d7ab63e2a6b3dacb03cb8#r120992574)). However, I think the same logic will also produce true for typical loads producing (non-internal) references into the GC heap. But because of the restriction on the opcodes of the input, `computeIsInternalPointer()` makes no claim about the result in such cases.